### PR TITLE
[general] Import/Export of Subnet/IpAddress

### LINKED ISSUE #18: [subnet] CSV Import and Export of subnets and their IPs

### PR DESCRIPTION
**Pull request by anurag-ks**
Implements and closes #18

### CODE REVIEW

**Review by nemesifier — COMMENTED** [review]
Is looking good but I need more details in order to test it and review it as explained in chat.

**Comment by anurag-ks** [comment]
@nemesisdesign I have sent an email regarding the details of testing on the ML.

File: .travis.yml [hunk 1]
```diff
@@ -21,9 +21,17 @@ install:
   - pip install "django>=2.0,<2.1"
   - python setup.py -q develop
 
+before_script:
+  # ensure no migrations pending but show output for easier debugging
```
Comment (nemesifier) [inline .travis.yml]: this change doesn't belong to this PR

File: django_ipam/base/models.py [hunk 2]
```diff
@@ -57,7 +60,7 @@ def request_ip(self, options=None):
 
 
 class AbstractIpAddress(TimeStampedEditableModel):
-    subnet = models.ForeignKey(swapper.get_model_name("django_ipam", "Subnet"),
+    subnet = models.ForeignKey('Subnet',
```
Comment (nemesifier) [inline django_ipam/base/models.py]: why are you doing this?

File: django_ipam/base/admin.py [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
 
 class AbstractSubnetAdmin(TimeReadonlyAdminMixin, ModelAdmin):
     change_form_template = "admin/django-ipam/subnet/change_form.html"
+    change_list_template = "admin/django-ipam/subnet/change_list.html"
+    app_name = "django_ipam"
 
     def change_view(self, request, object_id, form_url="", extra_context=None):
         instance = Subnet.objects.get(pk=object_id)
+        ipaddress_add_url = 'admin:{0}_ipaddress_add'.format(self.app_name)
+        ipaddress_change_url = 'admin:{0}_ipaddress_change'.format(self.app_name)
```
Comment (nemesifier) [inline django_ipam/base/admin.py]: is this for fixing the sample app? If yes move this change into a different patch from a different branch

File: django_ipam/base/models.py [hunk 4]
```diff
@@ -10,11 +12,12 @@
 
 
 class AbstractSubnet(TimeStampedEditableModel):
+    name = models.CharField(max_length=100, blank=True)
     subnet = NetworkField(db_index=True,
                           help_text=_('Subnet in CIDR notation, eg: "10.0.0.0/24" '
                                       'for IPv4 and "fdb6:21b:a477::9f7/64" for IPv6'))
     description = models.CharField(max_length=100, blank=True)
-    master_subnet = models.ForeignKey('self', on_delete=models.CASCADE,
+    master_subnet = models.ForeignKey('Subnet', on_delete=models.CASCADE,
```
Comment (nemesifier) [inline django_ipam/base/models.py]: why is this needed? is it for the sample app?

**Review by nemesifier — CHANGES_REQUESTED** [review]

File: django_ipam/templates/admin/django-ipam/subnet/change_form.html [hunk 5]
```diff
@@ -2,6 +2,16 @@
 {% load i18n admin_urls %}
 {% load static %}
 
+{% block object-tools-items %}
+{{ block.super }}
+<li>
+  <a href="{% url 'admin:export_ipaddress' subnet.id %}">Export Ip address</a>
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/change_form.html]: wrap this in `{% trans 'Export Ip address' %}`

File: django_ipam/templates/admin/django-ipam/subnet/change_form.html [hunk 6]
```diff
@@ -2,6 +2,16 @@
...
+</li>
+<li>
+  <a href="{% url 'admin:import_ipaddress' %}?_popup=1" onclick="return showAddAnotherPopup(this);">Import Ip address</a>
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/change_form.html]: same here, also put the link text on its own line so the code is more readable please

File: django_ipam/templates/admin/django-ipam/subnet/change_list.html [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{% extends "admin/change_list.html" %}
+{% load i18n admin_urls %}
+
+{% block object-tools-items %}
+{{ block.super }}
+<li>
+  <a href="{% url 'admin:import_ipaddress' %}?_popup=1" onclick="return showAddAnotherPopup(this);">Import Subnet/IP Address</a>
+</li>
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/change_list.html]: same here

File: django_ipam/templates/admin/django-ipam/subnet/import.html [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{% extends "admin/change_form.html" %}
+{% load i18n admin_urls %}
+
+{% block extrahead %}
+{{ block.super }}
+<title>Import IP address</title>
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/import.html]: wrap in trans tag

File: django_ipam/templates/admin/django-ipam/subnet/import.html [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+{% block breadcrumbs %}
+{% endblock %}
+
+{% block content %}
+  <div id="content-main">
+    <form action="{% url 'admin:import_ipaddress' %}" method="POST" enctype="multipart/form-data">
+      {% csrf_token %}
+      {{form.as_p}}
+      <div class="submit-row">
+        <input type="submit" value="Save" class="default" />
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/import.html]: wrap `Save` in trans tag

File: django_ipam/tests/base/test_admin.py [hunk 10]
```diff
@@ -104,3 +104,49 @@ def test_ipaddress_popup_response(self):
...
+        csv_data = """Sample Subnet\r
+        10.0.0.0/24\r
+        \r
+        ip_address,description\r
+        10.0.0.1,Testing\r
+        10.0.0.2,Testing\r
+        10.0.0.3,\r
+        10.0.0.4,\r
+        """
+
```
Comment (nemesifier) [inline django_ipam/tests/base/test_admin.py]: remove this blank line

File: django_ipam/tests/base/test_admin.py [hunk 11]
```diff
@@ -104,3 +104,49 @@ def test_ipaddress_popup_response(self):
...
+    def test_csv_upload(self):
+        csv_data = """Monachers - Matera,
+        10.27.1.0/24,
+        ,
+        ip address,description
+        10.27.1.1,Monachers
+        10.27.1.252,NanoStation M5
+        10.27.1.253,NanoStation M5
+        10.27.1.254,Nano Beam 5 19AC"""
+
```
Comment (nemesifier) [inline django_ipam/tests/base/test_admin.py]: remove this blank line

File: django_ipam/base/admin.py [hunk 12]
```diff
@@ -41,10 +46,43 @@ def change_view(self, request, object_id, form_url="", extra_context=None):
...
+                    self.export_ipaddress,
+                    name="export_ipaddress"),
+            path('import/', self.import_ipaddress, name="import_ipaddress"),
+        ]
+        return custom_urls + urls
+
+    def export_ipaddress(self, request, subnet_id):
+        response = HttpResponse(content_type='text/csv')
+        response['Content-Disposition'] = 'attachment; filename="ip_address.csv"'
+
```
Comment (nemesifier) [inline django_ipam/base/admin.py]: remove this blank line

File: django_ipam/base/models.py [hunk 13]
```diff
@@ -76,3 +79,29 @@ def clean(self):
...
+        writer.writerow(field.name for field in fields)
+        for obj in queryset:
+            row = []
+            for field in fields:
+                row.append(str(getattr(obj, field.name)))
+            writer.writerow(row)
+
+    def import_csv(self, file):
+        IpAddress = swapper.load_model("django_ipam", "IpAddress")
+        Subnet = swapper.load_model("django_ipam", "Subnet")
```
Comment (nemesifier) [inline django_ipam/base/models.py]: change to `subnet_model`

File: django_ipam/base/models.py [hunk 14]
```diff
@@ -76,3 +79,29 @@ def clean(self):
...
+        reader = csv.reader(StringIO(file.read().decode('utf-8')), delimiter=',')
```
Comment (nemesifier) [inline django_ipam/base/models.py]: this is a bit unreadable, break it at least on 2 lines.
Please document in the README the fact that the expected CSV delimiter is `,`

File: django_ipam/base/models.py [hunk 15]
```diff
@@ -76,3 +79,29 @@ def clean(self):
...
+        subnet = Subnet.objects.get_or_create(name=next(reader)[0].strip(), subnet=next(reader)[0].strip())[0]
```
Comment (nemesifier) [inline django_ipam/base/models.py]: this is unreadable, please break it on multiple line.
Please don't call `get_or_create` because it doesn't perform validation when creating objects.
You have to call `full_clean` and abort the operation gracefully if the validation fails.

File: django_ipam/base/models.py [hunk 16]
```diff
@@ -76,3 +79,29 @@ def clean(self):
...
+        next(reader)
+        next(reader)
+        for row in reader:
+            IpAddress.objects.get_or_create(subnet=subnet,
+                                            ip_address=row[0].strip(),
+                                            description=row[1].strip())
```
Comment (nemesifier) [inline django_ipam/base/models.py]: same here, you should do something like @rohithasrk has done in django-freeradius:

- check if ip exists, if does, continue
- create model instances and call `full_clean` but do not call `save` yet
- if validation fails, abort the operation
- otherwise put the model instance in a list
- once all instances have been validated, iterate the list again to save the instances

**Review by nemesifier — CHANGES_REQUESTED** [review]
I tried importing the CSV you sent on the ML and it works, but we need to improve this PR to make the result more professional, in the same way as we did with django-freeradius.

- move the import subnet button to the left, the add button should be on the far right
- do we have to use the pop up or can we open the page in the same window? I think same window would be better, some people hate popups and they have popup blockers
- please perform automatic conversion of XLS to CSV as discussed in the mailing list, you can use this example: https://stackoverflow.com/a/9884551/462707 - ensure we have a test for this case, include the XLS in the static files of the test app
- I tried supplying a PDF and the system crashed, you should manage the failure case
- please add a management command for importing and another one for exporting data and add this to the README
- add an API endpoint for importing and another one for exporting
- please include some basic info in the README in which you explain how to use this feature
- include a screenshot for the import button in the admin
- include a screenshot for the export button in the admin

File: README.rst [hunk 17]
```diff
@@ -355,13 +440,22 @@ The base API view classes can be extended into other django applications.
...
 
-    class SubnetView(BaseSubnetView):
+    class SubnetVew(BaseSubnetView):
         queryset = Subnet.objects.all()
 
 
     class IpAddressView(BaseIpAddressView):
         queryset = IpAddress.objects.all()
 
+    class ImportSubnetView(BaseImportSubnetView):
```
Comment (nemesifier) [inline README.rst]: you forgot one empty line here

File: django_ipam/base/admin.py [hunk 18]
```diff
@@ -47,11 +53,46 @@ def change_view(self, request, object_id, form_url="", extra_context=None):
...
+        form_template = 'admin/django-ipam/subnet/import.html'
+        if request.method == 'POST':
+            form = IpAddressImportForm(request.POST, request.FILES)
+            if form.is_valid():
+                file = request.FILES['csvfile']
+                if not file.name.endswith(('.csv', '.xls', '.xlsx')):
+                    messages.error(request, 'File type not supported.')
+                    return render(request, form_template, {'form': form})
+                if not Subnet.import_csv(self, file):
+                    messages.error(request, 'Import data error.')
```
Comment (nemesifier) [inline django_ipam/base/admin.py]: here you are not informing the user about what caused the error. This is not good because users need to know what they did wrong in order to fix it and complete the import process.
You should catch an exception here and get the details from the exception. I will explain below.

Keep in mind the` i18n` issue explained in the comment above.

File: django_ipam/base/models.py [hunk 19]
```diff
@@ -55,6 +59,70 @@ def request_ip(self, options=None):
...
+            reader = iter(row)
+        else:
+            reader = csv.reader(StringIO(file.read().decode('utf-8')),
+                                delimiter=',')
+        subnet_name = next(reader)[0].strip()
+        subnet_value = next(reader)[0].strip()
+        try:
+            subnet_exists = subnet_model.objects.filter(name=subnet_name, subnet=subnet_value).exists()
+        except (ValidationError, ValueError):
+            return False
```
Comment (nemesifier) [inline django_ipam/base/models.py]: First concern: this approach is not very pythonic.

In Python it is common to raise exceptions when something unexpected happens and then catch that exception anywhere it needs to be caught to handle it.

Example:
```python
class CsvImportException(Exception):
    pass
```

Then you can raise it with:

```python
raise CsvImportException(_('explain what went wrong here'))
```

When you catch it you can do:

```python
try:
    perform_import()  # pseudo-code
except CsvImportException as e:
    # e.message will contain the information about what went wrong
    # you have to show this users
    print(e.message)
```

**Second concern**

This code looks like it can be simplified, what do you think of the following:

```python
try:
    subnet = subnet_model.objects.get(name=subnet_name,
                                      subnet=subnet_value)
except subnet_model.DoesNotExist:
    try:
        subnet = subnet_model(name=subnet_name,
                              subnet=subnet_value)
        subnet.full_clean()
        subnet.save()
    except ValidationError as e:
        # here django should give us a human
        # readable error message which we can pass
        # to our exception,
        # but please verify this before doing it this way
        # there may be a better way
        message = str(e)
        raise CsvImportException(message)
    except ValueError as e:
        # why are we getting a ValueError?
        message = _('what went wrong and how to fix it?')
        raise CsvImportException(message)
```

File: django_ipam/base/models.py [hunk 20]
```diff
@@ -55,6 +59,70 @@ def request_ip(self, options=None):
...
+            if not ipaddress_model.objects.filter(subnet=subnet,
+                                                  ip_address=row[0].strip(),
+                                                  description=row[1].strip()).exists():
+                instance = ipaddress_model(subnet=subnet,
+                                           ip_address=row[0].strip(),
+                                           description=row[1].strip())
+                try:
+                    instance.full_clean()
+                except (ValidationError, ValueError):
+                    return False
```
Comment (nemesifier) [inline django_ipam/base/models.py]: same as before, raise the exception which is caught elsewhere to display the information about what went wrong

File: README.rst [hunk 21]
```diff
@@ -322,6 +322,91 @@ subnet         Subnet UUID
...
+++++++++++++++++++++
+
+Data can be imported from the admin interface by just clicking on the import button on the subnet view.
+
+.. image:: https://raw.githubusercontent.com/openwisp/django-ipam/master/docs/images/import.png
+
+Import file format
+++++++++++++++++++
+
+.. code-block:: text
```
Comment (nemesifier) [inline README.rst]: are you talking about the CSV here? Looks like it, but please make it explicit.

File: README.rst [hunk 22]
```diff
@@ -322,6 +322,91 @@ subnet         Subnet UUID
...
+.. code-block:: shell
+
+    ./manage.py import_subnet --file=<file path>
+
+From admin interface
+++++++++++++++++++++
+
+Data can be imported from the admin interface by just clicking on the import button on the subnet view.
+
+.. image:: https://raw.githubusercontent.com/openwisp/django-ipam/master/docs/images/import.png
```
Comment (nemesifier) [inline README.rst:395]: is this screenshot included in the PR?

File: README.rst [hunk 23]
```diff
@@ -322,6 +322,91 @@ subnet         Subnet UUID
...
+    ./manage.py export_subnet <subnet value>
+
+This would export the subnet if it exists on the database.
+
+From admin interface
+++++++++++++++++++++
+
+Data can be exported from the admin interface by just clicking on the export button on the subnet's admin change view.
+
+.. image:: https://raw.githubusercontent.com/openwisp/django-ipam/master/docs/images/export.png
```
Comment (nemesifier) [inline README.rst:375]: is this screenshot included in the PR?

File: django_ipam/templates/admin/django-ipam/subnet/change_form.html [hunk 24]
```diff
@@ -2,6 +2,20 @@
 {% load i18n admin_urls %}
 {% load static %}
 
+{% block object-tools-items %}
+{{ block.super }}
+<li>
+  <a href="{% url 'admin:export_ipaddress' subnet.id %}">
+    {% trans "Export Ip address" %}
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/change_form.html]: this wording is a bit misleading: we say "Export Ip address" but in the truth we are exporting the entire subnet.
I think it should be: "Export Data" or simply "Export".

File: django_ipam/templates/admin/django-ipam/subnet/change_form.html [hunk 25]
```diff
@@ -2,6 +2,20 @@
...
+  </a>
+</li>
+<li>
+  <a href="{% url 'admin:import_ipaddress' %}">
+    {% trans "Import Ip address" %}
```
Comment (nemesifier) [inline django_ipam/templates/admin/django-ipam/subnet/change_form.html]: Same here

File: django_ipam/management/commands/__init__.py [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+    def add_arguments(self, parser):
+        parser.add_argument('--file', type=str, help="File path to import csv data.")
+
+    def handle(self, *args, **options):
+        file = options['file']
+        try:
+            if not file.endswith(('.csv', '.xls', '.xlsx')):
+                raise CommandError('File type not supported')
+            with open(file, 'rb+') as csvfile:
+                if not self.subnet_model.import_csv(self, csvfile):
```
Comment (nemesifier) [inline django_ipam/management/commands/__init__.py]: as explained before, here you should catch our custom exception and show the details about what went wrong to the user

File: django_ipam/api/generics.py [hunk 27]
```diff
@@ -49,3 +52,30 @@ def post(self, request, *args, **kwargs):
...
+    serializer_class = ImportSubnetSerializer
+    authentication_classes = (SessionAuthentication, BasicAuthentication)
+    permission_classes = (DjangoModelPermissions,)
+
+    def post(self, request, *args, **kwargs):
+        file = request.FILES['csvfile']
+        if not file.name.endswith(('.csv', '.xls', '.xlsx')):
+            return Response('File type not supported.')
+        if not self.subnet_model.import_csv(self, file):
+            return Response('Import data error.')
```
Comment (nemesifier) [inline django_ipam/api/generics.py]: same here, you should display a bit more info to the user about what went wrong, I explained how to do this below.

Eg:

```python
try:
    self.subnet_model.import_csv(self, file)
catch CsvImportError as e:
    return Response({'error': e.message, status=400})
```

File: django_ipam/api/generics.py [hunk 28]
```diff
@@ -49,3 +52,30 @@ def post(self, request, *args, **kwargs):
...
+
+class BaseImportSubnetView(CreateAPIView):
+    serializer_class = ImportSubnetSerializer
+    authentication_classes = (SessionAuthentication, BasicAuthentication)
+    permission_classes = (DjangoModelPermissions,)
+
+    def post(self, request, *args, **kwargs):
+        file = request.FILES['csvfile']
+        if not file.name.endswith(('.csv', '.xls', '.xlsx')):
+            return Response('File type not supported.')
```
Comment (nemesifier) [inline django_ipam/api/generics.py]: should be:
```python
return Response({'error': _('File type not supported.'), status=400)
```

(don't remember if it's `status` or `status_code` please check)

File: django_ipam/api/generics.py [hunk 29]
```diff
@@ -49,3 +52,30 @@ def post(self, request, *args, **kwargs):
...
+        if not self.subnet_model.import_csv(self, file):
+            return Response('Import data error.')
+        return Response('Data imported successfully.')
```
Comment (nemesifier) [inline django_ipam/api/generics.py]:
```python
return Response({'detail': _('Data imported successfully.')})
```

File: django_ipam/base/admin.py [hunk 30]
```diff
@@ -47,11 +53,46 @@ def change_view(self, request, object_id, form_url="", extra_context=None):
...
+
+    def import_ipaddress(self, request):
+        form = IpAddressImportForm()
+        form_template = 'admin/django-ipam/subnet/import.html'
+        if request.method == 'POST':
+            form = IpAddressImportForm(request.POST, request.FILES)
+            if form.is_valid():
+                file = request.FILES['csvfile']
+                if not file.name.endswith(('.csv', '.xls', '.xlsx')):
+                    messages.error(request, 'File type not supported.')
```
Comment (nemesifier) [inline django_ipam/base/admin.py]: strings that are to be shown to users should always be `i18n` (internationalization) ready, eg: `_('File type not supported.')`

File: django_ipam/management/commands/__init__.py [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+                    raise CommandError('Import data error.')
```
Comment (nemesifier) [inline django_ipam/management/commands/__init__.py]: i18n issue

File: django_ipam/management/commands/__init__.py [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+
+    def add_arguments(self, parser):
+        parser.add_argument('subnet', type=str, help='Subnet to export.')
+
+    def handle(self, *args, **options):
+        subnet = options['subnet']
+        try:
+            instance = self.subnet_model.objects.get(subnet=ip_network(subnet))
+        except self.subnet_model.DoesNotExist:
+            raise CommandError('Subnet "%s" does not exist' % subnet)
```
Comment (nemesifier) [inline django_ipam/management/commands/__init__.py]: i18n issue

File: django_ipam/management/commands/__init__.py [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+        except ValueError:
+            raise CommandError("'%s' does not appear to be an IPv4 or IPv6 network" % subnet)
```
Comment (nemesifier) [inline django_ipam/management/commands/__init__.py]: i18n issue, find out how to use string interpolation in translatable strings here (look for "Python’s standard named-string interpolation syntax"):
https://docs.djangoproject.com/en/dev/topics/i18n/translation/#standard-translation

**Review by nemesifier — CHANGES_REQUESTED** [review]
Great improvements!

We are almost there. There are a few more things to improve before this can be given to the every-day user.

I left you some comments that should help you.

**Comment by nemesifier** [comment]
@anurag-ks: another thing I noticed is that the HTML <title> of the import page is empty, can you add a title there?

File: README.rst (hunk 22 above)
Comment (anurag-ks) [reply README.rst:395]: Yes, I have added the files, - https://github.com/openwisp/django-ipam/pull/31/files#diff-6b41e55da1de1a57e6145c5989ca21db

**Review by nemesifier — APPROVED** [review]
Good work, we'll be testing and improving this feature in the near future for sure. :+1:
